tensorflow/stream_executor/rocm/rocm_blas.cc

```diff
@@ -1438,8 +1439,38 @@ bool ROCMBlas::DoBlasGemm(
     float alpha, const DeviceMemory<Eigen::half> &a, int lda,
     const DeviceMemory<Eigen::half> &b, int ldb, float beta,
     DeviceMemory<Eigen::half> *c, int ldc) {
-  LOG(ERROR) << "fp16 sgemm is not implemented in this rocBLAS version";
-  return false;
+  VLOG(1) << port::Printf(
+      "doing rocBLAS SGEMM: at=%d bt=%d m=%llu n=%llu "
+      "k=%llu alpha=%f a=%p lda=%d b=%p ldb=%d beta=%f "
+      "c=%p ldc=%d",
+      static_cast<int>(transa), static_cast<int>(transb), m, n, k, alpha,
+      a.opaque(), lda, b.opaque(), ldb, beta, c->opaque(), ldc);
+  if (transa == blas::Transpose::kNoTranspose) {
+    if (lda < static_cast<int64>(m)) {
+      LOG(WARNING) << "GEMM lda was smaller than m (no transpose case); "
+                      "precondition violation";
+    }
+  } else {
+    if (lda < static_cast<int64>(k)) {
+      LOG(WARNING) << "GEMM lda (" << lda << ") was smaller than k (" << k
+                   << ") (transpose case); precondition violation";
+    }
+  }
+  if (transb == blas::Transpose::kNoTranspose) {
+    if (ldb < static_cast<int64>(k)) {
+      LOG(WARNING) << "GEMM ldb (" << ldb << ") was smaller than k (" << k
+                   << ") (no transpose case); precondition violation";
+    }
+  } else {
+    if (ldb < static_cast<int64>(n)) {
+      LOG(WARNING) << "GEMM ldb was smaller than n (transpose case); "
+                      "precondition violation";
+    }
+  }
+  return DoBlasInternal(
+      wrap::rocblas_hgemm, stream, true /* = pointer_mode_host */,
+      ROCMBlasTranspose(transa), ROCMBlasTranspose(transb), m, n, k, &alpha,
+      ROCMMemory(a), lda, ROCMMemory(b), ldb, &beta, ROCMMemoryMutable(c), ldc);
```

Comment:
@amcamd `&alpha` and `&beta` are of type `float` so this invocation is incorrect as `rocblas_hgemm` expects them be of type `rocblas_half*`.

However, looking at `rocBLAS` implementation it seems `rocblas_half` is defined as `uint16_t`. I'm wondering how should we properly populate `alpha` / `beta` in case they are originally in fp32 types?

Reply:
I think we need to simply convert `alpha`  from type `float` to type `Eigen::half` 
i.e. 
    `Eigen::half alpa_h(alpha);` 
and pass in `&alpha_h` instead of `&alpha`.

Same applies to beta.

The underlying assumption here is that the memory layout for `Eigen::half` matches that for `rocblas_half`. We are already making this assumoption for the pointers for `A`, `B` and `C`.

Reply:
unfortunately this doesn't work:

```
 error: cannot convert 'const Eigen::half*' to 'const rocblas_half* {aka const short unsigned int*}'
```

Reply:
that is strange...how come the same works for `A`, `B` and `C`? (granted they get passed via `ROCMMemory`, but that should return a `Eigen::half` pointer, if I am reading the template correctly)

Ignoring the why above,  the following then should work

```
Eigen::half alpa_h(alpha);
rocblas_half alpha_ushort = alpha_h.x;
```
and then pass `&alpha_ushort`

Reply:
@deven-amd I'm not saying `A,B,C` works yet. In fact I doubt they work. I just submitted this PR and ask CI to test :)

Reply:
ohh...then the fix here may require explicitly casting `Eigen::half*` as `rocblas_half*` for all pointers involved.  As I said before this is with the assumption that the memory layout matches...